File: python-service/app/services/aml_service.py
"""
Fuzzy-Matching RegTech AML & Sanction Screening Microservice.
Combines:
1. Jaro-Winkler string similarity
2. Levenshtein edit distance ratio
3. Soundex phonetic encoding
Against US Treasury OFAC Specially Designated Nationals (SDN) and UN Sanction Lists.
"""

from dataclasses import dataclass
from typing import List, Dict, Any, Optional

try:
    import jellyfish
except ImportError:
    jellyfish = None

# Curated OFAC SDN & UN Sanction benchmark database entities
BENCHMARK_SANCTIONED_ENTITIES = [
    "VLADIMIR PETROV",
    "SERGEI IVANOV",
    "MIKHAIL BOGDANOV",
    "DMITRY VOLKOV",
    "VIKTOR BOUT",
    "ALEXANDER LUKASHENKO",
    "KIM JONG UN",
    "BASHAR AL ASSAD",
    "QASEM SOLEIMANI",
    "AYMAN AL ZAWAHIRI",
    "HASSAN NASRALLAH",
    "ALI KHAMENEI",
    "NICOLAS MADURO",
    "TARECK EL AISSAMI",
    "ISMAIL HANIYEH",
    "YAHYA SINWAR",
    "MOHAMMAD DEIF",
    "ABU BAKR AL BAGHDADI",
    "OSAMA BIN LADEN",
    "ILHAM ALIYEV"
]

@dataclass
class AMLCheckResult:
    query_name: str
    matched_target: str
    risk_score: float
    decision: str  # 'PASS', 'MANUAL_REVIEW', 'REJECT'
    metrics: Dict[str, float]

class AMLSanctionEngine:
    def __init__(self, sanctioned_entities: Optional[List[str]] = None):
        self.sanctioned_list = sanctioned_entities or BENCHMARK_SANCTIONED_ENTITIES

    def _pure_soundex(self, s: str) -> str:
        """Pure Python fallback for American Soundex with W/V phonetic transliteration."""
        s = (s or "").upper().replace("W", "V")
        clean = [c for c in s if c.isalpha()]
        if not clean:
            return "0000"
        first = clean[0]
        mapping = {
            'B': '1', 'F': '1', 'P': '1', 'V': '1',
            'C': '2', 'G': '2', 'J': '2', 'K': '2', 'Q': '2', 'S': '2', 'X': '2', 'Z': '2',
            'D': '3', 'T': '3',
            'L': '4',
            'M': '5', 'N': '5',
            'R': '6'
        }
        coded = [first]
        prev = mapping.get(first, '0')
        for c in clean[1:]:
            code = mapping.get(c, '0')
            if code != '0' and code != prev:
                coded.append(code)
            prev = code
        out = ''.join(coded)
        return (out + "000")[:4]

    def _pure_levenshtein(self, s1: str, s2: str) -> int:
        """Pure Python fallback for Levenshtein edit distance."""
        if len(s1) < len(s2):
            return self._pure_levenshtein(s2, s1)
        if len(s2) == 0:
            return len(s1)
        previous_row = range(len(s2) + 1)
        for i, c1 in enumerate(s1):
            current_row = [i + 1]
            for j, c2 in enumerate(s2):
                insertions = previous_row[j + 1] + 1
                deletions = current_row[j] + 1
                substitutions = previous_row[j] + (c1 != c2)
                current_row.append(min(insertions, deletions, substitutions))
            previous_row = current_row
        return previous_row[-1]

    def _pure_jaro_winkler(self, s1: str, s2: str) -> float:
        """Pure Python Jaro-Winkler similarity."""
        if s1 == s2:
            return 1.0
        len1, len2 = len(s1), len(s2)
        if len1 == 0 or len2 == 0:
            return 0.0
        match_bound = max(len1, len2) // 2 - 1
        matches1 = [False] * len1
        matches2 = [False] * len2
        matches = 0
        for i in range(len1):
            start = max(0, i - match_bound)
            end = min(i + match_bound + 1, len2)
            for j in range(start, end):
                if matches2[j]:
                    continue
                if s1[i] == s2[j]:
                    matches1[i] = True
                    matches2[j] = True
                    matches += 1
                    break
        if matches == 0:
            return 0.0
        t = 0
        k = 0
        for i in range(len1):
            if not matches1[i]:
                continue
            while not matches2[k]:
                k += 1
            if s1[i] != s2[k]:
                t += 1
            k += 1
        t //= 2
        jaro = (matches / len1 + matches / len2 + (matches - t) / matches) / 3.0
        prefix = 0
        for i in range(min(4, min(len1, len2))):
            if s1[i] == s2[i]:
                prefix += 1
            else:
                break
        return jaro + prefix * 0.1 * (1.0 - jaro)

    def screen_name(self, input_name: str, threshold_reject: float = 0.85, threshold_review: float = 0.65) -> AMLCheckResult:
        normalized_input = (input_name or "").strip().upper()
        if not normalized_input:
            return AMLCheckResult(
                query_name="",
                matched_target="",
                risk_score=0.0,
                decision="PASS",
                metrics={"jaro_winkler": 0.0, "levenshtein_ratio": 1.0, "soundex_match": 0.0}
            )

        max_score = 0.0
        best_match = ""
        best_metrics = {}

        for entity in self.sanctioned_list:
            normalized_entity = entity.strip().upper()

            # 1. Jaro-Winkler
            if jellyfish:
                jw_score = jellyfish.jaro_winkler_similarity(normalized_input, normalized_entity)
                lev_dist = jellyfish.levenshtein_distance(normalized_input, normalized_entity)
                soundex_input = jellyfish.soundex(normalized_input)
                soundex_entity = jellyfish.soundex(normalized_entity)
            else:
                jw_score = self._pure_jaro_winkler(normalized_input, normalized_entity)
                lev_dist = self._pure_levenshtein(normalized_input, normalized_entity)
                soundex_input = self._pure_soundex(normalized_input)
                soundex_entity = self._pure_soundex(normalized_entity)

            # 2. Normalized Levenshtein ratio
            max_len = max(len(normalized_input), len(normalized_entity))
            lev_ratio = 1.0 - (lev_dist / max_len) if max_len > 0 else 1.0

            # 3. Soundex phonetic comparison
            soundex_match = 1.0 if soundex_input == soundex_entity else 0.0

            # Composite Weighted Risk Score: 50% JW + 35% Levenshtein + 15% Soundex
            composite_score = (0.50 * jw_score) + (0.35 * lev_ratio) + (0.15 * soundex_match)

            if composite_score > max_score:
                max_score = composite_score
                best_match = entity
                best_metrics = {
                    "jaro_winkler": round(jw_score, 4),
                    "levenshtein_ratio": round(lev_ratio, 4),
                    "soundex_match": soundex_match,
                    "soundex_code": soundex_input
                }

        final_risk = round(max_score, 4)
        if final_risk >= threshold_reject:
            decision = "REJECT"
        elif final_risk >= threshold_review:
            decision = "MANUAL_REVIEW"
        else:
            decision = "PASS"

        return AMLCheckResult(
            query_name=input_name,
            matched_target=best_match,
            risk_score=final_risk,
            decision=decision,
            metrics=best_metrics
        )

# Global Singleton
aml_sanction_engine = AMLSanctionEngine()

File: server/src/services/amlService.js
/**
 * RegTech Fuzzy-Matching AML & Sanction Screening Service
 * Evaluates sender/recipient names using:
 * 1. Jaro-Winkler distance
 * 2. Levenshtein edit distance ratio
 * 3. Soundex phonetic encoding
 * Against US Treasury OFAC SDN & UN Sanction lists.
 */

const axios = require('axios');
const FASTAPI_URL = process.env.FASTAPI_SERVICE_URL || 'http://localhost:8000';

const BENCHMARK_SANCTIONED_ENTITIES = [
  'VLADIMIR PETROV',
  'SERGEI IVANOV',
  'MIKHAIL BOGDANOV',
  'DMITRY VOLKOV',
  'VIKTOR BOUT',
  'ALEXANDER LUKASHENKO',
  'KIM JONG UN',
  'BASHAR AL ASSAD',
  'QASEM SOLEIMANI',
  'AYMAN AL ZAWAHIRI',
  'HASSAN NASRALLAH',
  'ALI KHAMENEI',
  'NICOLAS MADURO',
  'TARECK EL AISSAMI',
  'ISMAIL HANIYEH',
  'YAHYA SINWAR',
  'MOHAMMAD DEIF',
  'ABU BAKR AL BAGHDADI',
  'OSAMA BIN LADEN',
  'ILHAM ALIYEV'
];

class AMLService {
  // Pure JS Soundex implementation with W/V phonetic transliteration equivalence
  soundex(s) {
    if (!s) return '0000';
    // Normalize W -> V for multilingual phonetic matching (e.g. Wladimir -> Vladimir)
    const clean = s.toUpperCase().replace(/W/g, 'V').replace(/[^A-Z]/g, '');
    if (!clean.length) return '0000';

    const map = {
      B: '1', F: '1', P: '1', V: '1',
      C: '2', G: '2', J: '2', K: '2', Q: '2', S: '2', X: '2', Z: '2',
      D: '3', T: '3',
      L: '4',
      M: '5', N: '5',
      R: '6'
    };

    const first = clean[0];
    let coded = [first];
    let prev = map[first] || '0';

    for (let i = 1; i < clean.length; i++) {
      const code = map[clean[i]] || '0';
      if (code !== '0' && code !== prev) {
        coded.push(code);
      }
      prev = code;
    }

    return (coded.join('') + '000').substring(0, 4);
  }

  // Pure JS Levenshtein Distance
  levenshtein(s1, s2) {
    if (s1 === s2) return 0;
    if (!s1.length) return s2.length;
    if (!s2.length) return s1.length;

    const matrix = [];
    for (let i = 0; i <= s2.length; i++) {
      matrix[i] = [i];
    }
    for (let j = 0; j <= s1.length; j++) {
      matrix[0][j] = j;
    }

    for (let i = 1; i <= s2.length; i++) {
      for (let j = 1; j <= s1.length; j++) {
        if (s2.charAt(i - 1) === s1.charAt(j - 1)) {
          matrix[i][j] = matrix[i - 1][j - 1];
        } else {
          matrix[i][j] = Math.min(
            matrix[i - 1][j - 1] + 1, // substitution
            matrix[i][j - 1] + 1,     // insertion
            matrix[i - 1][j] + 1      // deletion
          );
        }
      }
    }

    return matrix[s2.length][s1.length];
  }

  // Pure JS Jaro-Winkler Similarity
  jaroWinkler(s1, s2) {
    if (s1 === s2) return 1.0;
    const len1 = s1.length;
    const len2 = s2.length;
    if (len1 === 0 || len2 === 0) return 0.0;

    const matchDistance = Math.floor(Math.max(len1, len2) / 2) - 1;
    const s1Matches = new Array(len1).fill(false);
    const s2Matches = new Array(len2).fill(false);
    let matches = 0;

    for (let i = 0; i < len1; i++) {
      const start = Math.max(0, i - matchDistance);
      const end = Math.min(i + matchDistance + 1, len2);
      for (let j = start; j < end; j++) {
        if (s2Matches[j]) continue;
        if (s1.charAt(i) === s2.charAt(j)) {
          s1Matches[i] = true;
          s2Matches[j] = true;
          matches++;
          break;
        }
      }
    }

    if (matches === 0) return 0.0;

    let t = 0;
    let point = 0;
    for (let i = 0; i < len1; i++) {
      if (!s1Matches[i]) continue;
      while (!s2Matches[point]) point++;
      if (s1.charAt(i) !== s2.charAt(point)) t++;
      point++;
    }
    t /= 2;

    const jaro = (matches / len1 + matches / len2 + (matches - t) / matches) / 3.0;
    let prefix = 0;
    for (let i = 0; i < Math.min(4, Math.min(len1, len2)); i++) {
      if (s1.charAt(i) === s2.charAt(i)) prefix++;
      else break;
    }

    return jaro + prefix * 0.1 * (1.0 - jaro);
  }

  // Fallback native evaluation
  screenLocally(inputName, thresholdReject = 0.85, thresholdReview = 0.65) {
    const normalizedInput = (inputName || '').trim().toUpperCase();
    if (!normalizedInput) {
      return {
        query_name: '',
        matched_target: '',
        risk_score: 0.0,
        decision: 'PASS',
        metrics: { jaro_winkler: 0.0, levenshtein_ratio: 1.0, soundex_match: 0.0 }
      };
    }

    let maxScore = 0.0;
    let bestMatch = '';
    let bestMetrics = {};

    for (const entity of BENCHMARK_SANCTIONED_ENTITIES) {
      const normalizedEntity = entity.trim().toUpperCase();
      const jw = this.jaroWinkler(normalizedInput, normalizedEntity);
      const levDist = this.levenshtein(normalizedInput, normalizedEntity);
      const maxLen = Math.max(normalizedInput.length, normalizedEntity.length);
      const levRatio = maxLen > 0 ? 1.0 - levDist / maxLen : 1.0;
      const soundexInput = this.soundex(normalizedInput);
      const soundexEntity = this.soundex(normalizedEntity);
      const soundexMatch = soundexInput === soundexEntity ? 1.0 : 0.0;

      // Composite Weighted Score: 50% JW + 35% Lev + 15% Soundex
      const composite = 0.50 * jw + 0.35 * levRatio + 0.15 * soundexMatch;

      if (composite > maxScore) {
        maxScore = composite;
        bestMatch = entity;
        bestMetrics = {
          jaro_winkler: Math.round(jw * 10000) / 10000,
          levenshtein_ratio: Math.round(levRatio * 10000) / 10000,
          soundex_match: soundexMatch,
          soundex_code: soundexInput
        };
      }
    }

    const finalRisk = Math.round(maxScore * 10000) / 10000;
    let decision = 'PASS';
    if (finalRisk >= thresholdReject) {
      decision = 'REJECT';
    } else if (finalRisk >= thresholdReview) {
      decision = 'MANUAL_REVIEW';
    }

    return {
      query_name: inputName,
      matched_target: bestMatch,
      risk_score: finalRisk,
      decision,
      metrics: bestMetrics
    };
  }

  // Primary API execution with resilient fallback
  async screenName(name, thresholdReject = 0.85, thresholdReview = 0.65) {
    try {
      const response = await axios.post(
        `${FASTAPI_URL}/compliance/screen`,
        { name, threshold_reject: thresholdReject, threshold_review: thresholdReview },
        { timeout: 800 }
      );
      if (response.data && response.data.success) {
        return response.data.data;
      }
    } catch (e) {
      // Fallback seamlessly to native JS algorithm
    }
    return this.screenLocally(name, thresholdReject, thresholdReview);
  }
}

module.exports = new AMLService();
